File: src/main/java/meowing/zen/utils/TickScheduler.java
package meowing.zen.utils;
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents;
import net.minecraft.client.MinecraftClient;
import java.util.PriorityQueue;
import java.util.Comparator;

public final class TickScheduler {
    private static final TickScheduler instance = new TickScheduler();
    private final PriorityQueue<ScheduledTask> taskQueue = new PriorityQueue<>(Comparator.comparingLong(ScheduledTask::executeTick));
    private long currentTick = 0;

    private record ScheduledTask(long executeTick, Runnable action) {}

    private TickScheduler() {}

    public static void schedule(long delayTicks, Runnable action) {
        instance.taskQueue.offer(new ScheduledTask(instance.currentTick + delayTicks, action));
    }

    public static void register() {
        ClientTickEvents.END_CLIENT_TICK.register(instance::onClientTick);
    }

    private void onClientTick(MinecraftClient client) {
        currentTick++;
        ScheduledTask task;
        while ((task = taskQueue.peek()) != null && currentTick >= task.executeTick()) taskQueue.poll().action().run();
    }
}

File: src/main/kotlin/meowing/zen/utils/TickScheduler.kt
package meowing.zen.utils

import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents
import net.minecraft.client.MinecraftClient
import java.util.PriorityQueue
import java.util.Comparator

object TickScheduler {
    private val taskQueue = PriorityQueue<ScheduledTask>(Comparator.comparingLong(ScheduledTask::executeTick))
    private var currentTick: Long = 0

    private data class ScheduledTask(val executeTick: Long, val action: Runnable)

    fun schedule(delayTicks: Long, action: Runnable) {
        taskQueue.offer(ScheduledTask(currentTick + delayTicks, action))
    }

    fun register() {
        ClientTickEvents.END_CLIENT_TICK.register(::onClientTick)
    }

    private fun onClientTick(client: MinecraftClient) {
        currentTick++
        var task: ScheduledTask?
        while (taskQueue.peek().also { task = it } != null && currentTick >= task!!.executeTick) {
            taskQueue.poll().action.run()
        }
    }
}